src/editor/codemirror/language-server/signatureHelp.ts

```diff
@@ -0,0 +1,171 @@
+import { Tooltip, showTooltip } from "@codemirror/tooltip";
+import { StateEffect, StateField, Text } from "@codemirror/state";
+import {
+  EditorView,
+  logException,
+  PluginValue,
+  ViewPlugin,
+  ViewUpdate,
+} from "@codemirror/view";
+import {
+  SignatureHelp,
+  SignatureHelpParams,
+  SignatureHelpRequest,
+  MarkupContent,
+} from "vscode-languageserver-protocol";
+import { offsetToPosition } from "./positions";
+import { BaseLanguageServerView } from "./common";
+
+interface SignatureChangeEffect {
+  pos: number;
+  result: SignatureHelp | null;
+}
+
+export const setSignatureHelpInformationEffect =
+  StateEffect.define<SignatureChangeEffect>({});
+
+interface SignatureHelpState {
+  // Do we need this?
+  pos: number;
+  tooltip: Tooltip | null;
+  result: SignatureHelp | null;
+}
+
+const signatureHelpTooltipField = StateField.define<SignatureHelpState>({
+  create: () => ({
+    pos: 0,
+    result: null,
+    tooltip: null,
+  }),
+  update(state, tr) {
+    for (const effect of tr.effects) {
+      if (effect.is(setSignatureHelpInformationEffect)) {
+        return updateSignatureHelpState(state, effect.value);
+      }
+    }
+    return state;
+  },
+  provide: (f) => showTooltip.from(f, (val) => val.tooltip),
+});
+
+export class SignatureHelpView
+  extends BaseLanguageServerView
+  implements PluginValue
+{
+  update({ docChanged, selectionSet, changes, transactions }: ViewUpdate) {
+    if (
+      (docChanged || selectionSet) &&
+      this.view.state.field(signatureHelpTooltipField).tooltip
+    ) {
+      this.triggerSignatureHelpRequest();
+    } else if (docChanged) {
+      const last = transactions[transactions.length - 1];
+      if (last.isUserEvent("input.type")) {
```

Comment:
I'm not very confident in this. Can we trigger it in weird scenarios? E.g. undo/redo?

Reply:
So far so good. Only merging this behind a flag to build confidence.